Weather DLC - Make it Rain Anything!

timothymtorres opened the pull request:
## About The Pull Request
##### Disclaimer - Some of the code/icons/sounds were ported from TMGC. 

Introducing more weather types! And yes, you can now have rain be reagent based!

<details>
<summary>Regular Rain</summary>

![dreamseeker_9S1WzdBW8Z](https://github.com/user-attachments/assets/83aaa1fe-9105-4e15-8455-0337c5c592ef)

</details>

<details>
<summary>Blood Rain</summary>

![dreamseeker_SZufAoXDFt](https://github.com/user-attachments/assets/209404dc-34dd-4381-9bab-9b84eaec2658)

</details>

<details>
<summary>Acid Rain</summary>

![dreamseeker_ngiOqdB5n2](https://github.com/user-attachments/assets/0ab953c6-a215-444a-bdbd-addfc2b1ddbb)

</details>

You can even make it rain ants, plasma, or drugs. All of their effects get applied to turfs, objects, and mobs depending on the weather options you select.

Did I mention... there is thunder?

<details>
<summary>Thunder Strikes</summary>

![Untitled video - Made with Clipchamp](https://github.com/user-attachments/assets/7c5c5930-6e0a-4706-a41b-3cbcc277bfd5)

</details>

<details>
<summary>Sand Storms</summary>

![dreamseeker_ZEFUS73dSA](https://github.com/user-attachments/assets/4a754f2d-c4e5-4b2f-9add-4742628cfa74)

</details>

Despite all this new stuff, none of it has been directly added to the game but the code can be used in the future for:
- Wizard event - Similar to lava on floor, but this time make the reagent random or picked from a list and give wizard immunity
- Chaplain ability - Maybe make this a benefit or ability once enough favor has been obtained
- Admin events - I have added a BUNCH of admin tooling to run customized weather events that let you control a lot of options
- New station maps/biomes for downstreams (Jungle Station, etc.)
- Change Ion storms to use the new weather system that triggers EMP/thunder effects across the station
- IceBox could get plasma rain
- Lavaland could get thunder effects applied to ash storms

Relevant PRs that removed/added some of the weather stuff I used:
- #60303
- #25222

---

#### Attribution
- Rain sprites were added via [novaepee](https://github.com/novaepee) in https://github.com/tgstation/TerraGov-Marine-Corps/pull/9675
- Sand sprites were added via [TiviPlus](https://github.com/TiviPlus) (who commissioned them from bimmer) in https://github.com/tgstation/TerraGov-Marine-Corps/pull/4645
- Rain sounds [already existed on tg](https://github.com/tgstation/tgstation/pull/25222#discussion_r106794579) and were provided by provided by Cuboos, using Royalty Free sounds, presumed under default tg sound license - Creative Commons 3.0 BY-SA
- Siren sound is from siren.wav by IFartInUrGeneralDirection -- [Freesound](https://freesound.org/s/46092/) -- License: Attribution 4.0

The original `siren.ogg` sound used on a lot of SS13 servers doesn't seem to have any attribution that I could locate. The sound was added about 15 years ago. So I just looked for a somewhat similar sounding siren noise on Freesound.

## Why It's Good For The Game
More weather customization!

## Changelog
:cl:
add: Added new weather types for rain and sandstorms. Rain now uses a reagent that gets exposed to the turfs, mobs, and objects. There is also a thunder strike setting you can apply to any weather.
add: Hydro trays and soil will now add reagents to their trays when exposed to a chemical reaction. (weather, shower, chem spills, foam grenades, etc.)
add: Weather temperature now affects weather reagents and mobs body temperature.
bal: Snowstorm temperature calculations were tweaked to allow universal weather temperature effects.
sound: Added weather sounds from TGMC for rain and sirens (attributed to Cuboos and IFartInUrGeneralDirection )
image: Added weather images from TGMC for rain and sand storms (attributed to Novaepee and Bimmer)
refactor: Refactored a lot of the weather code to be more robust
admin: Admins can now control more weather related options when running weather events. The weather admin verbs have been moved to their own "Weather" category under the Admin tab.
/:cl:

timothymtorres committed:
Add rain storms to weather subsystem

timothymtorres committed:
Add greyscale rain

timothymtorres committed:
Add rain messages by reagent name and more rain subtypes

timothymtorres committed:
Fix rain messages

timothymtorres committed:
Fix rainstorm defines

timothymtorres committed:
Add greyscale to all rain animations

timothymtorres committed:
Add global list for rain storm sounds

timothymtorres committed:
Add  weather listener element to object possession for rain storms

timothymtorres committed:
Apply Melbert's weather fix

timothymtorres committed:
Fix deprecated code comment about sound loop

timothymtorres committed:
Add rain looping sound

timothymtorres committed:
Add rain stuff

timothymtorres committed:
Add weather flags

timothymtorres committed:
Add weather flags to weather subsystem

timothymtorres committed:
Replace boolean var with weather flag for WEATHER_BAROMETER

timothymtorres committed:
Remove old deprecated code

timothymtorres committed:
Change rain looping sound into 3 sections

timothymtorres committed:
More weather flags and weather_act procs

timothymtorres committed:
Change low rain icon to have less rain sprites

timothymtorres committed:
Tweak code comments for bitflag

timothymtorres committed:
Add weather flags to storms

timothymtorres committed:
Tweak weather vars

timothymtorres committed:
Add thunder weather effects

timothymtorres committed:
Add IS_WEATHER_AESTHETIC macro

timothymtorres committed:
Add WEATHER_ENDLESS bitflags to storms

timothymtorres committed:
Add weather bitflags to inherit parent types

timothymtorres committed:
Add sand storm weather events

timothymtorres committed:
Fix weather_listener ZTRAITS

timothymtorres committed:
Repath /datum/weather/rain to be /datum/weather/rain_storm

timothymtorres committed:
Add immunity traits for sand storms and rain storms

timothymtorres committed:
Remove deprecated code and code comments

timothymtorres committed:
Add weather immunity traits to trait lists

timothymtorres committed:
Add weather_flags to bitfield defines

timothymtorres committed:
Add admin verb support to run custom reagents on rain weather

timothymtorres committed:
Code cleanup

timothymtorres committed:
Move weather admin verbs to their own cateogry

grungussuss reviewed (changes requested):
These all need attribution. Include a .txt file in the same folder as the files themselves, ideally from the original source, not just the repo you ported it from.

optimumtact commented:
![image](https://github.com/user-attachments/assets/8f730c4c-50a3-4bea-a572-d6d905bdb3a8)

timothymtorres committed:
Optimize weather turf processing to skip openspace

timothymtorres committed:
Switch the order since walls are more common

timothymtorres committed:
Merge branch 'master' into port-tgmc-weather

timothymtorres committed:
Fix linter being a cry baby

timothymtorres committed:
Add attributions

timothymtorres committed:
Replace unknown siren.ogg with Freesound siren.ogg version

Wallemations commented:
Does this only affect certain areas or is it just a blanket z-level thing?

timothymtorres commented:
> Does this only affect certain areas or is it just a blanket z-level thing?

The weather subsystem gives you a lot of control over selecting location:
- Select by z-level although this can result in multi-z if using a station trait like `ZTRAIT_SNOWSTORM`
- Affected areas list determines which areas your weather will target (ie. `/area/station/chapel`)
- Protected areas list determines which areas get ignored (ie. `/area/station/maintenance`)
- Indoor bitflag determines if the weather will affect an area marked as outdoors (ie. `/area/space` is outside, but `/area/station` is inside)

All of these get combined to basically filter out areas that should only be impacted.

Loafin34 commented:
Moses McMosesFace: "HoS, Let my assistants go, or the station will suffer pestilence and plague."
Pharoah Hosmuntep: "You have no power here."
_blood rain_
https://www.youtube.com/watch?v=g98eFTYZugs

timothymtorres committed:
Reagent color is applied directly

timothymtorres committed:
More cleanup

timothymtorres committed:
Merge branch 'master' into port-tgmc-weather

timothymtorres committed:
Fix duplicate code from merge

timothymtorres committed:
Optimize weather subsystem processing

timothymtorres committed:
Refactor the weather processing to be more robust

timothymtorres commented:
Two minor things left that I need to finish:
- Fix admin triggered weather events ignoring weather looping sounds (technically a bug that existed before this PR)
- Improve admin customization options for weather events (friendler UI)

The rest of the code is in a reviewable state @LemonInTheDark . I added your recommended changes from the discord for speed improvements.

timothymtorres committed:
Code comment docstring

timothymtorres committed:
Fix iteration bug with ++

timothymtorres committed:
Fix list looping with modulo divison for REAL this time

timothymtorres committed:
Fix if loop only allowing one subsystem task

timothymtorres committed:
Add weather notification bitflag for mob messages

timothymtorres committed:
Remove code comments

timothymtorres committed:
Fix rain reagent type

timothymtorres committed:
Convert fractions for thunder and weather turfs to be a prob() chance to increase the number by +1

timothymtorres committed:
Optomize weather processing to ignore space turfs for thunder/reagent effects

timothymtorres committed:
Fix fraction conversion for prob()

timothymtorres committed:
Filter out fulltile windows from weather/thunder turfs

timothymtorres committed:
Add some useful thunder chance defines

timothymtorres committed:
Change insane thunder define to be slightly more insane

timothymtorres committed:
Optimize our weather turfs by skipping fulltile windows

timothymtorres committed:
Add extra desc to thunder defines

timothymtorres committed:
Add tgui stuff for admin weather events

timothymtorres committed:
Fix argument for thunder not overriding due to null check

timothymtorres committed:
Add weather reagents get added to hydroponics plant trays

timothymtorres committed:
Merge branch 'master' into port-tgmc-weather

timothymtorres committed:
Fix admin weather ignoring thunder and custom reagent

EricZilla commented:
Mummy themed heretic with a sandstorm would go pretty hard

timothymtorres committed:
Improve lighting and sound to be more robust

timothymtorres committed:
Remove hydro tray rain handling

timothymtorres commented:
All done now ✅

timothymtorres committed:
Add time defines

LemonInTheDark commented on code/controllers/subsystem/weather.dm:
```diff
@@ -8,24 +8,66 @@ SUBSYSTEM_DEF(weather)
+			while(current_mobs_cache.len)
+				var/mob/living/target = current_mobs_cache[current_mobs_cache.len]
+				current_mobs_cache.len--
+				if(QDELETED(target))
+					continue
```
this shouldn't be possible, if we delete a mob it should get removed from your copied list too (as in, you should go muck around in living destroy)

timothymtorres replied:
I'll remove it then. Pretty sure I c/p'd this code from another proc so maybe I'll check if it's necessary there as well.

LemonInTheDark replied:
for clarity the mob being deleted needs to be handled, I just think it should be handled in the mob's destroy, rather then just letting the reference hang like this.

timothymtorres replied:
So if the weather subsystem lags out and in between ticks a mob gets `QDELETED` but they would still be inside our old copied player list. (since we cached it) The list gets reset every tick so in the next subsystem tick it will use a new player list without the `QDELETED` mob. 

It makes sense to simple skip the mob here and let the list reset. Are you worried about hard deletes or something? Do I need to switch it to `WEAKREFS()`

LemonInTheDark commented on code/controllers/subsystem/weather.dm:
```diff
@@ -8,24 +8,66 @@ SUBSYSTEM_DEF(weather)
+
+		if(weather_event.currentpart == SSWEATHER_TURFS)
+			if(!resumed)
+				weather_event.turf_iteration = floor(weather_event.weather_turfs_per_tick)
+				var/fraction_chance = fract(weather_event.weather_turfs_per_tick) * 100
+				weather_event.turf_iteration += prob(fraction_chance) ? 1 : 0
```
what's goin on here?

timothymtorres replied:
We are using the `fract` function to convert the remaining decimal from `turfs_per_tick` into an integer based on `prob()`.

If `weather_turfs_per_tick = 45.25` then that remaining `0.25` gets put into `prob()` so we can convert it to a +1 if it succeeds, otherwise it's a 0. Why don't we do this automatically with something like `ceil()`, `round()`, or `floor()`?

I already tried that. The problem is when you need to handle really low chances for small areas. If say, an admin wants to run the blood rain weather with thunder in a place like chapel that only has 50 turfs, the thunder chance is generally really tiny... like `0.0001` so when you multiply you get a `turfs_per_tick` of like `0.0005`. Any rounding, ceiling, or flooring operations here make it guaranteed that weather effects for the turf will be either nothing or extreme. You do not want thunder happening on 1/50 tiles because it will be insane.

That's why we need to use `prob()` with our decimal. It keeps things consistent. `45.25` means you have 45 turfs that will be affected and then a 25% chance of an extra turf also being affected. In our case of `0.0005` it means you have a 5% chance a turf being affected.

Hopefully that makes sense.

LemonInTheDark replied:
I see. you could make it more concise but I think this is fine. I see what you're after, you're trying to handle two different cases at once.

timothymtorres replied:
I rewrote it to be apart of the math defines:
```dm
/// Probabilistic rounding: Adds 1 to the integer part of x with a probability equal to the decimal part of x.
/// ie. ROUND_PROB(40.25) returns 40 with 75% probability, and 41 with 25% probability.
#define ROUND_PROB(x) ( floor(x) + (prob(fract(x) * 100)) )
```

Applied via:
- https://github.com/tgstation/tgstation/pull/89550/commits/b510e0d9775bc61a270e185dbae3d94eba5fc2f6
- https://github.com/tgstation/tgstation/pull/89550/commits/537e6c0a82e915e1c3351987cf6c639bdb19ecf1

LemonInTheDark commented on code/controllers/subsystem/weather.dm:
```diff
@@ -8,24 +8,66 @@ SUBSYSTEM_DEF(weather)
+
+		if(weather_event.currentpart == SSWEATHER_THUNDER)
+			if(!resumed)
+				weather_event.thunder_iteration = floor(weather_event.thunder_turfs_per_tick)
+				var/fraction_chance = fract(weather_event.thunder_turfs_per_tick) * 100
+				weather_event.thunder_iteration += prob(fraction_chance) ? 1 : 0
```
similar

timothymtorres replied:
Same reasoning as above. If you want me to I can make this a macro somewhere since I think it's a useful but niche operation.

LemonInTheDark commented on code/datums/weather/weather.dm:
```diff
@@ -288,3 +392,9 @@
 
 	return gen_overlay_cache
+
+/// Updates the currentpart with the subsystem task that is next in line
+/datum/weather/proc/next_subsystem_task()
+	// loops back to the start of the list once it reaches the end
+	var/next_part = currentpart % length(subsystem_tasks) + 1
+	currentpart = subsystem_tasks[next_part]
```
should just use WRAP(), saves you the comment

LemonInTheDark replied:
also, the name implies it's a getter, not a proc that assumes it's called in a particular context that then also sets a var. should rename it, or just return the list access, makes things a bit less musty (low priority concern)

timothymtorres replied:
I was looking for a proc that would do this but I couldn't find anything. I'm going to delete the `next_subsystem_task()` and just use the macro instead.

timothymtorres replied:
Replaced with `WRAP_UP` via https://github.com/tgstation/tgstation/pull/89550/commits/a793e7d9fc129a90941e5127330fb1334d5fb4a1

LemonInTheDark commented on code/controllers/subsystem/weather.dm:
```diff
@@ -8,24 +8,66 @@ SUBSYSTEM_DEF(weather)
 		eligible_zlevels -= z
-		var/randTime = rand(3000, 6000)
-		next_hit_by_zlevel["[z]"] = addtimer(CALLBACK(src, PROC_REF(make_eligible), z, possible_weather), randTime + initial(our_event.weather_duration_upper), TIMER_UNIQUE|TIMER_STOPPABLE) //Around 5-10 minutes between weathers
+		var/randTime = rand(5 MINUTES, 10 MINUTES)
```
❤️

timothymtorres replied:
Small recognitions like this are nice for any code reviews because they are grueling. Ty

LemonInTheDark commented on code/controllers/subsystem/weather.dm:
```diff
@@ -49,7 +91,7 @@ SUBSYSTEM_DEF(weather)
 			eligible_zlevels["[z]"][weather] = probability
 
-/datum/controller/subsystem/weather/proc/run_weather(datum/weather/weather_datum_type, z_levels)
+/datum/controller/subsystem/weather/proc/run_weather(datum/weather/weather_datum_type, z_levels, area_type, weather_bitflags, thunder_value, datum/reagent/reagent_type)
```
that's a LOTTTT

timothymtorres replied:
This is needed so admins can pass arguments to the weather when they use their custom weather verb. The bitflags shrunk it down significantly.

timothymtorres replied:
I'm going to make the arguments here into a named list

timothymtorres replied:
Applied via https://github.com/tgstation/tgstation/pull/89550/commits/1c854922b9edef2dad3a53532ad6778aa2394618

LemonInTheDark commented on code/__DEFINES/weather.dm:
```diff
@@ -1,9 +1,54 @@
+	"None - Admin Safe Space (Thunder Disabled)" = NONE,
+))
+
+/// Does weather have any type of processing related to mobs, turfs, or thunder?
+#define IS_WEATHER_AESTHETIC(flags) (!(flags & (WEATHER_TURFS|WEATHER_MOBS|WEATHER_THUNDER)))
```
why have this macro take flags instead of a type/why have it as a macro instead of a proc?

timothymtorres replied:
We don't want type checking here because admins verbs allow custom weather flags to be changed. And I don't have any issue making it a proc, I just thought the macro was faster.

LemonInTheDark replied:
ok so like, if this is your performance bottleneck we're FUCKED
for clarity I meant like, `#define IS_WEATHER_AESTHETIC(weather_datum) (!weather_datum.flags...` but I think that's a bit redundant on further thought. I think this is a little awkward and we're not tryin to paper over hell behavior, so I think it'd be better to just be clear.

What I think you should do is just make a bitfield that is "non asthetic weather" and use that as a field instead of abstracting it away like this (for similar reasons to why we don't use wrapper macros for bitflag ops).

so like
```
#define FUNCTIONAL_WEATHER (WEATHER_TURFS|WEATHER_MOBS|WEATHER_THUNDER)
....

if(!(weather_datum.flags & FUNCTIONAL_WEATHER))
```
and such.

If you don't agree we me we can yell at each other bout it that's fine, I just think the abstraction here is giving you less then is really helpful.

timothymtorres replied:
Applied via https://github.com/tgstation/tgstation/pull/89550/commits/27e9d1ad77096fdebfb0e9ff3c770a9003b2ce48

LemonInTheDark commented on code/datums/weather/weather.dm:
```diff
@@ -83,18 +79,63 @@
+	area_type = area_override || area_type
+	weather_flags = weather_flags_override || weather_flags
+
+	// turf_thunder_chance = thunder_chance_override || turf_thunder_chance
+	// this breaks when thunder_chance_override is 0 (aka FALSE), so we need to null check
```
??

timothymtorres replied:
I was adding a code comment on why this argument override doesn't work like the others.

```dm
/obj/foo
  bar = 5

/obj/foo/New(value_override)
  bar = value_override || bar

var/obj/foo/foo_1 = new(21)
world.log << foo_1.bar // 21

var/obj/foo/foo_2 = new(0)
world.log << foo_2.bar // 5 (when it should be 0)
```
I was just listing a reason why we do the null check here and can definitely remove it if needed.

LemonInTheDark replied:
it's not like hurting anything but I didn't really understand it, I think it's a little redundant

timothymtorres replied:
Removed via 2c90307b494425e399132af7a571fa7284cc8db6

LemonInTheDark commented on code/datums/weather/weather.dm:
```diff
@@ -105,28 +146,68 @@
+		send_alert(telegraph_message, telegraph_sound, telegraph_sound_vol)
+	addtimer(CALLBACK(src, PROC_REF(start)), telegraph_duration)
+
+/datum/weather/proc/setup_weather_areas(list/selected_areas)
+	if(length(selected_areas))
+		return // impacted areas already been setup
+
```
I'm sorry this pattern stinks please just return a list of selected areas pass by ref is HELL

timothymtorres replied:
I'll see if I can rework it.

timothymtorres replied:
This was refactored and removed via https://github.com/tgstation/tgstation/commit/05e0edb485b6a57d294186401c4435308e1782d4

LemonInTheDark commented on code/datums/weather/weather.dm:
```diff
@@ -105,28 +146,68 @@
-	if(telegraph_duration)
-		send_alert(telegraph_message, telegraph_sound, telegraph_sound_vol)
-	addtimer(CALLBACK(src, PROC_REF(start)), telegraph_duration)
+/datum/weather/proc/setup_weather_turfs()
+	for(var/area/weather_area as anything in impacted_areas)
+		for(var/z in impacted_z_levels)
+			for(var/turf/valid_weather_turf as anything in weather_area.get_turfs_by_zlevel(z))
+				// applying weather effects to solid walls is a waste since nothing will happen
+				if(isclosedturf(valid_weather_turf))
+					continue
+				// same logic for space and openspace turfs which should boost performance a ton
+				// note - mobs in space/openspace turfs still have weather affects applied to them if they are in a affected area
+				if(is_space_or_openspace(valid_weather_turf))
+					continue
+				// solid windows are also worth skipping
+				var/obj/structure/window/window = locate() in valid_weather_turf
+				if(window?.fulltile)
+					continue
+
+				weather_turfs += valid_weather_turf
```
this code makes me really anxious, this list is going to be really long and we're doin a lot of extra work here, how expensive is this on lavaland (I understand that's where you want to use this)

LemonInTheDark replied:
(also we can't make any guarantees about the state of these turfs AFTER the weather starts which will make this unreliable as a list of open turfs)

LemonInTheDark replied:
basically, why aren't we just getting random valid turfs from the area instead of doing this? (also unrelated you should add dupe protection to turf picking code

timothymtorres replied:
I haven't profiled any of the code yet. 

I can try to maybe make a weighted list for the areas where the weights are determined by the total turfs in an area. That might work and allow areas to be updated dynamically while our weather is running.

LemonInTheDark replied:
please do profile, also yeah that'd be a good idea. Lavaland is HUGE and this is a blocking operation so I'm apprehensive. I don't want the "weather started" lag to come back.

timothymtorres replied:
Refactored turf selection code to use a weighted list of the area/z-level. It is setup when the weather event starts and is not modified during the weather at all. While testing I ran into a runtime where I started the weather event while the arrival shuttle was docked and then when the weather started it tried to pick a turf from the arrival shuttle (since that area was in the weighted list initially) but the list was empty since the shuttle moved.

Since this is a rare edge case (and areas rarely change) I made it skip a picked area if it doesn't return a turf.

It is possible to run:
```dm
/datum/weather/setup_weather_areas() // updates which areas are apart of impacted areas
/datum/weather/setup_weather_turfs() // recalculates number of turfs to process
/datum/weather/update_areas() // updates overlays of impacted areas
```
every time the weather event is fired. I don't think these procs are expensive but I think it's wasted time because areas updating their size/location is a rare occurrence. I also want to get the Tracy byond profiler working before I do any real profiling since the native DM one is pretty bare bones.

Applied changes via https://github.com/tgstation/tgstation/pull/89550/commits/05e0edb485b6a57d294186401c4435308e1782d4

LemonInTheDark commented on code/datums/weather/weather.dm:
```diff
@@ -105,28 +146,68 @@
+
+	var/total_turfs = length(weather_turfs)
+
+	if(!total_turfs || !(weather_flags & (WEATHER_TURFS|WEATHER_THUNDER)))
+		return
```
why do we return if it's 0 doesn't that mean the var SHOULD be 0?

timothymtorres replied:
Both `weather_turfs_per_tick` and `thunder_turfs_per_tick` default to 0 so they don't need to be set if there are no valid turfs in the weather areas. (which can happen if it's all walled off or space)

LemonInTheDark replied:
oook

LemonInTheDark commented on code/datums/weather/weather_types/rain_storm.dm:
```diff
@@ -0,0 +1,153 @@
+	setup_weather_areas(impacted_areas)
+
+	GLOB.rain_storm_sounds.Cut()
+	for(var/area/impacted_area as anything in impacted_areas)
+		GLOB.rain_storm_sounds[impacted_area] = /datum/looping_sound/rain/start
+		CHECK_TICK
+	return ..()
+
+/datum/weather/rain_storm/start()
+	GLOB.rain_storm_sounds.Cut()
+	for(var/area/impacted_area as anything in impacted_areas)
+		GLOB.rain_storm_sounds[impacted_area] = /datum/looping_sound/rain/middle
+		CHECK_TICK
+	return ..()
+
+/datum/weather/rain_storm/wind_down()
+	GLOB.rain_storm_sounds.Cut()
+	for(var/area/impacted_area as anything in impacted_areas)
+		GLOB.rain_storm_sounds[impacted_area] = /datum/looping_sound/rain/end
+		CHECK_TICK
+	return ..()
```
should be no reason to sleep here this is not an expensive operation

timothymtorres replied:
Sure

timothymtorres replied:
Applied via 8c11299c43c9194b4ff4d0899f3f2e69fa8fec18

LemonInTheDark commented on code/datums/weather/weather_types/rain_storm.dm:
```diff
@@ -0,0 +1,153 @@
+	rain_reagent.expose_mob(living, TOUCH, RAIN_REAGENT_VOLUME)
+
+	var/rain_type = LOWER_TEXT(rain_reagent.name)
+	if((weather_flags & WEATHER_NOTIFICATION) && prob(5))
```
early return

timothymtorres replied:
Yep

timothymtorres replied:
Applied via 5975d7689f505e5f3beb7c360652c52e7a012941

LemonInTheDark replied:
what happen to this commit?

timothymtorres replied:
5975d7689f505e5f3beb7c360652c52e7a012941

timothymtorres replied:
Oh wait. I see what happened. I copied the commit ID and posted the response here before pushing to remote. So Github didn't see the commit hash.

LemonInTheDark commented on code/datums/weather/weather_types/rain_storm.dm:
```diff
@@ -0,0 +1,153 @@
+
+/datum/weather/rain_storm/weather_act_turf(turf/open/weather_turf)
+	for(var/obj/thing as anything in weather_turf.contents)
+		if(!thing.IsObscured())
```
early continue

timothymtorres replied:
Agree

timothymtorres replied:
Applied via e242c070e4c524fefdca1bc46007063ab2235089

LemonInTheDark commented on code/datums/weather/weather_types/rain_storm.dm:
```diff
@@ -0,0 +1,153 @@
+
+	/// A weighted list of possible reagents that will rain down from the sky.
+	/// Only one of these will be selected to be used as the reagent
+	var/list/whitelist_weather_reagents = list(/datum/reagent/water)
```
this isn't weighted tho

timothymtorres replied:
This list is used with `pick_weight_recursive()` that accepts regular lists, nested lists, and weighted lists. Other rain subtypes do use weighted lists. (see `/datum/weather/rain_storm/acid`)

timothymtorres replied:
Tweaked code comments to include it via bf298f94d1dc6953d1ea265f46eb0a09d7889892

LemonInTheDark commented on code/datums/weather/weather_types/sand_storm.dm:
```diff
@@ -0,0 +1,68 @@
+
+	weather_flags = (WEATHER_MOBS | WEATHER_BAROMETER | WEATHER_NOTIFICATION)
+
+/datum/weather/sand_storm/telegraph()
+	GLOB.sand_storm_sounds.Cut()
+	for(var/area/impacted_area as anything in impacted_areas)
+		GLOB.sand_storm_sounds[impacted_area] = /datum/looping_sound/weak_outside_ashstorm
+		CHECK_TICK
+
+	return ..()
+
+/datum/weather/sand_storm/start()
+	GLOB.sand_storm_sounds.Cut()
+	for(var/area/impacted_area as anything in impacted_areas)
+		GLOB.sand_storm_sounds[impacted_area] = /datum/looping_sound/active_outside_ashstorm
+		CHECK_TICK
+	return ..()
+
+/datum/weather/sand_storm/wind_down()
+	GLOB.sand_storm_sounds.Cut()
+	for(var/area/impacted_area as anything in impacted_areas)
+		GLOB.sand_storm_sounds[impacted_area] = /datum/looping_sound/weak_outside_ashstorm
+		CHECK_TICK
+	return ..()
```
similar

timothymtorres replied:
Agree

timothymtorres replied:
Applied via 20e92de23c82045e2da6eff19483d1d84cbe457a

LemonInTheDark commented on code/datums/weather/weather_types/snow_storm.dm:
```diff
@@ -21,19 +21,18 @@
 	end_sound_vol = /datum/looping_sound/snowstorm::volume + 10
 
 	area_type = /area
-	protect_indoors = TRUE
```
?

timothymtorres replied:
It was converted to a `WEATHER_INDOORS` bitflag for the `weather_flags`

LemonInTheDark replied:
the snowstorm datum doesn't use that flag is what I meant, sorry

timothymtorres replied:
Ah I see. The bitflag was inverted which is why it's not included.

var/protect_indoors = TRUE // Weather can NOT affect indoor areas
weather_bitflag = WEATHER_INDOORS // Weather can affect indoor areas

LemonInTheDark replied:
ah I see

LemonInTheDark commented on code/modules/admin/verbs/adminweather.dm:
```diff
@@ -0,0 +1,78 @@
+ADMIN_VERB(run_weather, R_ADMIN|R_FUN, "Run Weather", "Triggers specific weather on the z-level you choose.", ADMIN_CATEGORY_WEATHER)
+
+	var/static/list/weather_choices = list()
+	if(!length(weather_choices))
+		for(var/datum/weather/weather_type as anything in subtypesof(/datum/weather))
+			weather_choices[initial(weather_type.type)] = weather_type
```
s no reason for this to be static really the list is gonna be real short

timothymtorres replied:
Agree

timothymtorres replied:
Applied via bb83bf3c67b0f4da41885f8e3fa59ba7056fa051

LemonInTheDark commented on code/modules/admin/verbs/adminweather.dm:
```diff
@@ -0,0 +1,78 @@
+			return
+		if("Cancel")
+			return
+		//if("Custom") venture on lil fella
```
?

timothymtorres replied:
Just wanted to acknowledge that the `Custom` choice is being selected here instead of omitting it.

timothymtorres replied:
Removed via https://github.com/tgstation/tgstation/pull/89550/commits/e15582c23655129578c394f5aa8556a4b2495f3c

LemonInTheDark commented on code/modules/admin/verbs/adminweather.dm:
```diff
@@ -0,0 +1,78 @@
+			return
+		//if("Custom") venture on lil fella
+
+	var/static/list/area_choices = list()
+	if(!length(area_choices))
+		for(var/area/area_type as anything in typesof(/area))
+			area_choices[initial(area_type.type)] = area_type
```
similar to some extent, unless this is really choking in which case I take it back

timothymtorres replied:
Remove the static list?

LemonInTheDark replied:
I mean like, it's not a huge deal either way, it prob isn't saving you that much time here is all

timothymtorres replied:
Applied via d065486eaa83251eb51707d0d600229ca89f2b23

LemonInTheDark commented on code/modules/admin/verbs/adminweather.dm:
```diff
@@ -0,0 +1,78 @@
+		return
+	area_choice = area_choices[area_choice]
+
+	var/weather_bitflags = input_bitfield(
+		user,
+		"Weather flags - Select the flags for your weather event",
+		"weather_flags",
+		weather_choice::weather_flags,
+	)
+
```
shouldn't these come from the type?

timothymtorres replied:
Yes, it's whichever weather type the admin selects. This allows you to edit the bitflags to toggle things like weather affecting mobs, turfs, thunder, etc.

LemonInTheDark commented on code/modules/admin/verbs/adminweather.dm:
```diff
@@ -0,0 +1,78 @@
+	)
+
+	var/datum/reagent/reagent_choice
+	if(ispath(weather_choice, /datum/weather/rain_storm) && (weather_bitflags & (WEATHER_TURFS|WEATHER_MOBS)))
```
why the explicit typecheck?

timothymtorres replied:
So I allowed something like thunder to be used by ANY weather subtypes because I can see it being used for ashstorms, sandstorms, etc. In fact one of the other maintainers suggested thunder with ashstroms on lavaland.

The reagent stuff was really only meant to be used by rain. None of the other weather types are greyscaled like rain to inherit reagent color for their sprites.

Could I disable this check to allow reagents to be used for other weather events? Sure, but it just doesn't make sense. Acid snow? (but the snow is white) A holywater sandstorm?

So I just restricted it to the rain weather subtypes.

LemonInTheDark replied:
it seems a bit strange conceptually because rain isn't the only weather effect that could in theory have reagents (it's ugly w that setup proc if we're implying generic behavior it should be generic)

I think if you wanted bespoke setup it should happen in an overridable proc on the datum itself (though you would need to change the setup process some).

LemonInTheDark reviewed (changes requested):
are we actually using these new weather types/behaviors for anything? am a bit reticent to add a bunch of dead code

timothymtorres commented:
I responded to most of your code reviews. Planning on making the code changes tomorrow.

> are we actually using these new weather types/behaviors for anything? am a bit reticent to add a bunch of dead code

I already have a new wizard event setup for magical rain in my fork at - https://github.com/timothymtorres/tgstation/tree/wizard-rain-event

Just waiting for this PR to get reviewed and merged before I make a PR for it.

LemonInTheDark commented:
what about the sandstorm?

LemonInTheDark commented on code/datums/weather/weather.dm:
```diff
@@ -83,18 +79,63 @@
+
+	// turf_thunder_chance = thunder_chance_override || turf_thunder_chance
+	// this breaks when thunder_chance_override is 0 (aka FALSE), so we need to null check
+	turf_thunder_chance = !isnull(thunder_chance_override) ? thunder_chance_override : turf_thunder_chance
```
```suggestion
	turf_thunder_chance = isnull(thunder_chance_override) ? turf_thunder_chance : thunder_chance_override 
```
something something double negative. that or just like make this an if else that'd also be fine

LemonInTheDark commented on code/datums/weather/weather.dm:
```diff
@@ -83,18 +79,63 @@
+	// this breaks when thunder_chance_override is 0 (aka FALSE), so we need to null check
+	turf_thunder_chance = !isnull(thunder_chance_override) ? thunder_chance_override : turf_thunder_chance
+
+	if(IS_WEATHER_AESTHETIC(weather_flags))
+		return
+
+	if(weather_flags & (WEATHER_MOBS))
+		subsystem_tasks += SSWEATHER_MOBS
+	if(weather_flags & (WEATHER_TURFS))
+		subsystem_tasks += SSWEATHER_TURFS
+	if(weather_flags & (WEATHER_THUNDER))
+		subsystem_tasks += SSWEATHER_THUNDER
 
+	currentpart = subsystem_tasks[1]
```
```suggestion
	if(weather_flags & (WEATHER_MOBS))
		subsystem_tasks += SSWEATHER_MOBS
	if(weather_flags & (WEATHER_TURFS))
		subsystem_tasks += SSWEATHER_TURFS
	if(weather_flags & (WEATHER_THUNDER))
		subsystem_tasks += SSWEATHER_THUNDER

	if(length(subsystem_tasks))
		currentpart = subsystem_tasks[1]
```
dealer's choice, I think this is a bit more clear

LemonInTheDark commented on code/controllers/subsystem/weather.dm:
```diff
@@ -8,24 +8,66 @@ SUBSYSTEM_DEF(weather)
-		var/datum/weather/our_event = V
-		if(our_event.aesthetic || our_event.stage != MAIN_STAGE)
+	for(var/datum/weather/weather_event in processing)
+		if(IS_WEATHER_AESTHETIC(weather_event.weather_flags) || weather_event.stage != MAIN_STAGE)
 			continue
```
similar here why not check if the subsystem has any tasks it's asking for?

timothymtorres replied:
They both are accomplishing the same thing. (since subsystems get determined via bitflags)

Applied via https://github.com/tgstation/tgstation/pull/89550/commits/af5f9c2bf5358b1730476e1504acf648f2fe639a

LemonInTheDark commented on code/datums/weather/weather_types/rain_storm.dm:
```diff
@@ -0,0 +1,153 @@
+/datum/weather/rain_storm/New(z_levels, area_override, weather_flags_override, thunder_chance_override, datum/reagent/custom_reagent)
+	..()
+
+	if(IS_WEATHER_AESTHETIC(weather_flags))
+		return
+
+	var/reagent_id
```
are we checking for anything here or just the turf effects flag

timothymtorres replied:
If the weather doesn't process on turfs, mobs, or thunder then no need to setup reagent for our weather event. (since nothing is gonna happen) Although doing it this way disables any selected reagent from being used to color, which might be something we want to keep for only aesthetics.

timothymtorres replied:
Removed the early return via https://github.com/tgstation/tgstation/pull/89550/commits/f69b183951435c1bede58ba7e3de91c3982c0c42

timothymtorres committed:
Refactor subsystem handling

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

timothymtorres committed:
Remove double negative

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

timothymtorres committed:
Make list regular

timothymtorres committed:
Merge branch 'port-tgmc-weather' of https://github.com/timothymtorres/tgstation into port-tgmc-weather

timothymtorres committed:
Remove another static

timothymtorres committed:
Remove check tick code fom sandstorms

timothymtorres committed:
Add early continue statement

timothymtorres committed:
Add early return for weather notification

timothymtorres committed:
Remove sleep from rain_storm

timothymtorres committed:
Add code comment for lists

timothymtorres committed:
Remove deprecated code comment

timothymtorres committed:
Replace increment subsystem tasks with

timothymtorres committed:
Fix weather not wrapping

timothymtorres committed:
Add ROUND_PROB() math function

timothymtorres committed:
Clarify the code comments

timothymtorres committed:
Refactor bitflag macro to use chained bitflags

timothymtorres committed:
Remove early return to allow reagent to color weather

timothymtorres committed:
Remove commented out code

timothymtorres committed:
Replace bitflag check with subsystem length check

timothymtorres committed:
Refactor weather subsystem to use weighted area/z-level lists to pick turfs

timothymtorres committed:
Fix linter

timothymtorres committed:
Optimize weather setup areas

timothymtorres committed:
Refactor weather args to use a list

timothymtorres committed:
Change rain reagents to apply to any weather subtypes

timothymtorres committed:
Fix missing parenthesis

timothymtorres committed:
Fix typo

timothymtorres committed:
Fix other weather subtypes not using inheritance

timothymtorres committed:
Fix turfs ignoring weather reagents

timothymtorres commented:
Addressed all your review concerns with changes. Still need feedback on what you want to do with `if(QDELETED(target))`. 

> what about the sandstorm? I'm not like ontologically opposed to it I just don't want dead code in the repo so much as I can help it, so if we aren't doing anything w it it shouldn't exist. (TBH I'd almost prefer you merged this kind of refactor with the tied feature)

We have biome generation code already which will compliment weather nicely. There are also a lot of extra pieces of components and objs that are exclusively admin spawned. (ex. beetlejuice component) I choose not to add the features bundled in with this PR because this is a rather large overhaul of a subsystem. Including several weather related features that is going to affect balance IMO is going to get buried in this PR and overlooked. 

I can definitely add the wizard rain and thunder to lavaland as a bare minimum if you want not a big deal either way.

deathrobotpunch commented:
it’s raining hydrochloric acid wow oooof my skin

timothymtorres committed:
Change admin verb message

timothymtorres committed:
Allow hydroponic trays to be affected by the weather

timothymtorres committed:
Refactor weather to use reagent holder and expose() proc

timothymtorres commented:
Some small updates:
- Refactored weather to use a nullspace reagent holder that allows a lot more flexibility. (multiple reagents, temperature, etc.)
- Weather reagents apply their effects via `expose()` method which triggers a lot more reactions and signals without skipping stuff
- Weather reagents can now be added to hydro trays and soil due to the refactor
- Adjusted weather reagent exposure: 15u for turfs/objects (applied via RNG), 5u for mobs (applied every tick). This volume difference compensates for the reduced frequency on turfs/objects.

It's now possible to stick buckets and barrels outside to collect rain. 

![giphy](https://github.com/user-attachments/assets/642b0f9f-c6a2-4559-8fde-b48027e15f12)

timothymtorres committed:
Merge branch 'master' into port-tgmc-weather

timothymtorres committed:
Merge branch 'master' into port-tgmc-weather

timothymtorres committed:
Move turf reagent volume multiplier to weather defines

timothymtorres committed:
Change weather reagents and mob bodytemperature to be affected by weather temperature

timothymtorres committed:
Change snowstorms to be colder than default icebox atmos

timothymtorres committed:
Fix weather subsystem not wrapping properly

timothymtorres committed:
Add deep fried rain weather event

TheVekter commented:
TMed on all servers.

SmArtKar commented on code/modules/admin/verbs/adminweather.dm:
```diff
@@ -0,0 +1,90 @@
+	var/custom_choice = tgui_alert(user, "How would you like to run the weather settings?", "Custom Weather", custom_options)
+	switch(custom_choice)
+		if("Default")
+			SSweather.run_weather(weather_choice, z_level) // default settings
```
Doesn't log that admin used this verb, needs fixing - your logging code is right at the bottom

timothymtorres replied:
Applied via https://github.com/tgstation/tgstation/pull/89550/commits/8fcf3c1a6af197d680b26181dfbc31e4c20d9e41

LemonInTheDark commented:
Does this pr make snow happen on all zs of icebox? I'm seeing it happen on live and I cannot work out why it's occuring

LemonInTheDark commented:
FUCKING RUINS

timothymtorres commented:
> Does this pr make snow happen on all zs of icebox? I'm seeing it happen on live and I cannot work out why it's occuring

I have not changed the rendering code or how z-levels or weather areas are selected. My turf selection code for thunder and weather reagents uses the existing area setup code.

This has been a known bug that predates this weather PR:
- #88392

My best guess is that multi-z hiccups pretty bad with certain planes. See:
- #53783
- #71757

timothymtorres committed:
Add logging to admin default weather event

timothymtorres committed:
Change probability for rainstorms and sandstorms so they will actually spawn if used as a map trait

timothymtorres committed:
Small optimization

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

timothymtorres committed:
Change deep fried weather to be raining oil

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

timothymtorres committed:
Add early return to setup weather

timothymtorres commented on code/datums/weather/weather.dm:
```diff
@@ -105,28 +185,69 @@
 		for(var/z in impacted_z_levels)
-			if(length(affected_area.turfs_by_zlevel) >= z && length(affected_area.turfs_by_zlevel[z]))
+			var/total_turfs = length(affected_area.turfs_by_zlevel) >= z && length(affected_area.turfs_by_zlevel[z])
+			if(total_turfs)
```
Applied via https://github.com/tgstation/tgstation/pull/89550/commits/70d2443583b9eae5eb811383cc9aa8436d3a80d8

timothymtorres commented on code/datums/weather/weather.dm:
```diff
@@ -224,16 +344,115 @@
+
+	var/reagent_name = LOWER_TEXT(weather_reagent.name)
+	var/weather_name = replacetext(name, " storm", "") // Remove " storm" if it exists
+	var/wetmessage = pick(
```
If the suit gives protection then they won't get the message.

```dm
if((immunity_type && HAS_TRAIT(mob_to_check, immunity_type)) || HAS_TRAIT(mob_to_check, TRAIT_WEATHER_IMMUNE))
	return
```

There is already a notification flag for the name of the reagent that gets early returned. It is a few lines above this code comment.

```dm
if(!(weather_flags & WEATHER_NOTIFICATION) || prob(95))
	return
```

timothymtorres replied:
Removed via https://github.com/tgstation/tgstation/pull/89550/commits/eee59f2c62a11f26598b9af1fd30761d893015c3

timothymtorres committed:
Remove admin weather category

timothymtorres commented on code/__DEFINES/admin_verb.dm:
```diff
@@ -81,6 +81,7 @@ _ADMIN_VERB(verb_path_name, verb_permissions, verb_name, verb_desc, verb_categor
 #define ADMIN_CATEGORY_FUN "Admin.Fun"
 #define ADMIN_CATEGORY_GAME "Admin.Game"
 #define ADMIN_CATEGORY_SHUTTLE "Admin.Shuttle"
+#define ADMIN_CATEGORY_WEATHER "Admin.Weather"
```
Applied via https://github.com/tgstation/tgstation/pull/89550/commits/5181b37f584d02d4ed68e74f597aaa8f6532e8fb

timothymtorres committed:
Remove weather rain reagent notifications
